File: typescript/packages/metadata/src/core/file-source.ts
// FileSource — a MetaDataSource backed by a file on disk. Server-side only
// (touches node:fs); lives under src/core/ so the browser-safe root never
// imports it.

import { basename, extname } from "node:path";
import type { MetaDataFormat, MetaDataSource } from "../loader/meta-data-source.js";

/** Infer a source format from a file extension. `.yaml`/`.yml` → "yaml";
 *  everything else (including `.json`) → "json", the canonical default. */
function inferFormat(path: string): MetaDataFormat {
  const ext = extname(path).toLowerCase();
  return ext === ".yaml" || ext === ".yml" ? "yaml" : "json";
}

// File reader — Bun-first (Bun.file().text()) with a Node fallback
// (node:fs/promises.readFile). Matches CLAUDE.md's Bun-first / Node-compatible policy.
let _readText: ((path: string) => Promise<string>) | undefined;

async function getReadText(): Promise<(path: string) => Promise<string>> {
  if (_readText !== undefined) return _readText;
  if (typeof Bun !== "undefined") {
    _readText = (p) => Bun.file(p).text();
  } else {
    const { readFile } = await import("node:fs/promises");
    _readText = (p) => readFile(p, "utf-8");
  }
  return _readText;
}

/** A metadata source backed by a file on disk. */
export class FileSource implements MetaDataSource {
  readonly id: string;
  readonly format: MetaDataFormat;
  private readonly _path: string;

  constructor(path: string) {
    this._path = path;
    // basename() for readable error messages; cross-platform (handles both / and \). Full path retained for read().
    this.id = basename(path);
    this.format = inferFormat(path);
  }

  /** The absolute/relative path this source reads from. */
  get path(): string {
    return this._path;
  }

  async read(): Promise<string> {
    const readText = await getReadText();
    return readText(this._path);
  }
}

File: server/csharp/MetaObjects/Loader/FileSource.cs
// FileSource — a metadata source backed by a file on disk.
//
// Ported from typescript/packages/metadata/src/loader/core/file-source.ts

namespace MetaObjects.Loader;

/// <summary>
/// A metadata source backed by a file on disk. The format is inferred from the
/// file extension: <c>.yaml</c> / <c>.yml</c> → <see cref="MetaDataFormat.Yaml"/>,
/// everything else → <see cref="MetaDataFormat.Json"/>.
/// </summary>
public sealed class FileSource : IMetaDataSource
{
    /// <summary>The full path to the backing file.</summary>
    public string FilePath { get; }

    public string Id { get; }

    public MetaDataFormat Format { get; }

    public FileSource(string path)
    {
        FilePath = path;
        Id = System.IO.Path.GetFileName(path);
        Format = MetaDataFormats.InferFromExtension(path);
    }

    /// <summary>
    /// Read the file content. BOM stripping is handled by the parser, not here.
    /// </summary>
    public string Read() => System.IO.File.ReadAllText(FilePath);
}
